src/Data/Foreign/Generic/Enum.purs

```diff
@@ -0,0 +1,105 @@
+  :: forall a rep
+   . Generic a rep
+  => GenericDecodeEnum rep
+  => Options
+  -> Foreign
+  -> F a
+genericDecodeEnum opts = map to <<< decodeEnum opts
+
+genericEncodeEnum
+  :: forall a rep
+   . Generic a rep
+  => GenericEncodeEnum rep
+  => Options
+  -> a
+  -> Foreign
+genericEncodeEnum opts = encodeEnum opts <<< from
+
+class GenericDecodeEnum a where
+  decodeEnum :: Options -> Foreign -> F a
+
+class GenericEncodeEnum a where
+  encodeEnum :: Options -> a -> Foreign
+
+instance sumGenericDecodeEnum
+  :: (GenericDecodeEnum a, GenericDecodeEnum b)
+  => GenericDecodeEnum (Sum a b) where
+  decodeEnum opts f = Inl <$> decodeEnum opts f <|> Inr <$> decodeEnum opts f
+
+instance ctorNoArgsGenericDecodeEnum
+  :: IsSymbol name
+  => GenericDecodeEnum (Constructor name NoArguments) where
+  decodeEnum {sumEncoding: TaggedObject {constructorTagTransform}} f = do
+    tag <- readString f
+    unless (constructorTagTransform tag == ctorName) $
+      fail (ForeignError ("Expected " <> show ctorName <> " tag for unary constructor literal " <> ctorName))
+    pure $ Constructor NoArguments
+    where
+      ctorName = constructorTagTransform $ reflectSymbol (SProxy :: SProxy name)
+
+instance ctorArgumentGenericDecodeEnum
+  :: Fail """
+    genericDecodeEnum cannot be used with constructors with arguments, as it is meant to work with string literal "Enums", Sum types of no-argument constructors.
+  """
+  => GenericDecodeEnum (Constructor name (Argument a)) where
+  decodeEnum _ _ = unsafeCrashWith "unreachable decodeEnum was reached."
+
+instance ctorProductGenericDecodeEnum
+  :: Fail """
+    genericDecodeEnum cannot be used with Product types, as it is meant to work with string literal "Enums", Sum types of no-argument constructors.
+  """
+  => GenericDecodeEnum (Constructor name (Product a b)) where
+  decodeEnum _ _ = unsafeCrashWith "unreachable decodeEnum was reached."
+
+instance ctorRecGenericDecodeEnum
+  :: Fail """
+    genericDecodeEnum cannot be used with Record types, as it is meant to work with string literal "Enums", Sum types of no-argument constructors.
+  """
+  => GenericDecodeEnum (Constructor name (Rec a)) where
+  decodeEnum _ _ = unsafeCrashWith "unreachable decodeEnum was reached."
+
+instance sumGenericEncodeEnum
+  :: (GenericEncodeEnum a, GenericEncodeEnum b)
+  => GenericEncodeEnum (Sum a b) where
+  encodeEnum opts (Inl a) = encodeEnum opts a
+  encodeEnum opts (Inr b) = encodeEnum opts b
+
+instance ctorNoArgsGenericEncodeEnum
+  :: IsSymbol name
+  => GenericEncodeEnum (Constructor name NoArguments) where
+  encodeEnum {sumEncoding: TaggedObject {constructorTagTransform}} _ = toForeign ctorName
+    where
+      ctorName = constructorTagTransform $ reflectSymbol (SProxy :: SProxy name)
+
+instance ctorArgumentGenericEncodeEnum
+  :: Fail """
+    genericEncodeEnum cannot be used with constructors with arguments, as it is meant to work with string literal "Enums", Sum types of no-argument constructors.
```

Comment:
I would still trim this down a bit and put the stuff about enums into the documentation for the classes.